File: prono-core/frontend/src/api/newsletter.ts
import apiClient from './axios';

export type NewsletterTheme = 'FOOTBALL' | 'F1';
export type NewsletterStatus = 'DRAFT' | 'SENT';

export interface Newsletter {
  id: number;
  title: string;
  subtitle?: string;
  bodyMd: string;
  theme: NewsletterTheme;
  ctaLabel?: string;
  ctaUrl?: string;
  status: NewsletterStatus;
  sentCount: number;
  createdBy?: string;
  createdAt?: string;
  sentAt?: string;
}

export interface NewsletterInput {
  title: string;
  subtitle?: string;
  bodyMd: string;
  theme: NewsletterTheme;
  ctaLabel?: string;
  ctaUrl?: string;
}

export const listNewsletters = async (): Promise<Newsletter[]> => {
  const res = await apiClient.get<Newsletter[]>('/admin/newsletter');
  return res.data;
};

export const createNewsletter = async (input: NewsletterInput): Promise<Newsletter> => {
  const res = await apiClient.post<Newsletter>('/admin/newsletter', input);
  return res.data;
};

export const updateNewsletter = async (id: number, input: NewsletterInput): Promise<Newsletter> => {
  const res = await apiClient.put<Newsletter>(`/admin/newsletter/${id}`, input);
  return res.data;
};

export const deleteNewsletter = async (id: number): Promise<void> => {
  await apiClient.delete(`/admin/newsletter/${id}`);
};

export const previewNewsletter = async (id: number): Promise<string> => {
  const res = await apiClient.get<string>(`/admin/newsletter/${id}/preview`, { responseType: 'text' });
  return res.data;
};

export const sendNewsletterTest = async (id: number, targetEmail: string): Promise<void> => {
  await apiClient.post(`/admin/newsletter/${id}/test`, { targetEmail });
};

export const broadcastNewsletter = async (id: number): Promise<number> => {
  const res = await apiClient.post<{ recipientCount: number }>(`/admin/newsletter/${id}/send`);
  return res.data.recipientCount;
};

File: prono-core/backend/src/main/java/com/pronocore/entity/Newsletter.java
package com.pronocore.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;

/**
 * A one-shot broadcast campaign (e.g. announcing a big feature shipped to prod).
 * Unlike the transactional templates in {@code service.email.template}, the body
 * is authored by an admin (Markdown) and persisted. A campaign is edited as a
 * DRAFT then broadcast once, after which it is frozen as SENT.
 */
@Entity
@Table(name = "newsletter")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Newsletter {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(length = 200)
    private String subtitle;

    @Column(name = "body_md", nullable = false, columnDefinition = "TEXT")
    private String bodyMd;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private Theme theme = Theme.FOOTBALL;

    @Column(name = "cta_label", length = 100)
    private String ctaLabel;

    @Column(name = "cta_url", length = 500)
    private String ctaUrl;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private Status status = Status.DRAFT;

    @Column(name = "sent_count", nullable = false)
    @Builder.Default
    private int sentCount = 0;

    @Column(name = "created_by", length = 100)
    private String createdBy;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "sent_at")
    private LocalDateTime sentAt;

    public enum Status {
        DRAFT, SENT
    }

    public enum Theme {
        FOOTBALL, F1
    }
}
